snapshot/backup.go

```diff
@@ -358,75 +357,85 @@ func (snap *Builder) chunkify(imp importer.Importer, record *importer.ScanRecord
 
 	objectHasher, releaseGlobalHasher := snap.repository.GetPooledMACHasher()
 	defer releaseGlobalHasher()
-
-	var firstChunk = true
-	var cdcOffset uint64
 	var object_t32 objects.MAC
 
 	var totalEntropy float64
 	var totalFreq [256]float64
 	var totalDataSize uint64
 
 	// Helper function to process a chunk
-	processChunk := func(data []byte) error {
+	processChunk := func(idx int, data []byte) error {
 		var chunk_t32 objects.MAC
 
-		chunkHasher, releaseChunkHasher := snap.repository.GetPooledMACHasher()
-		if firstChunk {
+		if idx == 0 {
 			if object.ContentType == "" {
 				object.ContentType = mimetype.Detect(data).String()
 			}
-			firstChunk = false
 		}
 
-		chunkHasher.Write(data)
-		copy(chunk_t32[:], chunkHasher.Sum(nil))
-		releaseChunkHasher()
-
 		entropyScore, freq := entropy(data)
 		if len(data) > 0 {
-			for i := 0; i < 256; i++ {
+			for i := range 256 {
 				totalFreq[i] += freq[i]
 			}
 		}
-		chunk := objects.NewChunk()
-		chunk.ContentMAC = chunk_t32
-		chunk.Length = uint32(len(data))
-		chunk.Entropy = entropyScore
 
-		object.Chunks = append(object.Chunks, *chunk)
-		cdcOffset += uint64(len(data))
+		chunkHasher, releaseChunkHasher := snap.repository.GetPooledMACHasher()
+		chunkHasher.Write(data)
+		copy(chunk_t32[:], chunkHasher.Sum(nil))
+		releaseChunkHasher()
 
-		totalEntropy += chunk.Entropy * float64(len(data))
+		object.Chunks[idx].ContentMAC = chunk_t32
+		object.Chunks[idx].Entropy = entropyScore
+
+		totalEntropy += object.Chunks[idx].Entropy * float64(len(data))
 		totalDataSize += uint64(len(data))
 
-		return snap.repository.PutBlobIfNotExists(resources.RT_CHUNK, chunk.ContentMAC, data)
+		return snap.repository.PutBlobIfNotExists(resources.RT_CHUNK, object.Chunks[idx].ContentMAC, data)
 	}
 
 	if record.FileInfo.Size() == 0 {
 		empty := []byte{}
-		// Produce an empty chunk for empty file
 		objectHasher.Write(empty)
-		if err := processChunk(empty); err != nil {
+
+		chunk := objects.NewChunk()
+		object.Chunks = append(object.Chunks, *chunk)
+
+		if err := processChunk(0, empty); err != nil {
 			return nil, err
 		}
 	} else if record.FileInfo.Size() < int64(snap.repository.Configuration().Chunking.MinSize) {
-		// Small file case: read entire file into memory
 		cdcChunk, err := io.ReadAll(rd)
 		if err != nil {
 			return nil, err
 		}
 		objectHasher.Write(cdcChunk)
-		if err := processChunk(cdcChunk); err != nil {
+
+		chunk := objects.NewChunk()
+		chunk.Length = uint32(len(cdcChunk))
+		object.Chunks = append(object.Chunks, *chunk)
+
+		if err := processChunk(0, cdcChunk); err != nil {
 			return nil, err
 		}
 	} else {
+		semaphoreSize := record.FileInfo.Size() / int64(snap.repository.Configuration().Chunking.MinSize)
+		if semaphoreSize < 1 {
+			semaphoreSize = 1
+		}
+		if semaphoreSize > int64(backupCtx.maxConcurrency) {
+			semaphoreSize = int64(backupCtx.maxConcurrency)
+		}
+		semaphore := make(chan struct{}, 42)
```

Comment:
the answer to life, the universe and everything, but maybe we should use `semaphoreSize` :P